src/nvim/window.c

```diff
@@ -4196,26 +4199,35 @@ int win_new_tabpage(int after, char *filename)
 
     newtp->tp_topframe = topframe;
     last_status(false);
-
     redraw_all_later(UPD_NOT_VALID);
 
-    tabpage_check_windows(old_curtab);
+    if (enter) {
+      tabpage_check_windows(old_curtab);
 
-    lastused_tabpage = old_curtab;
+      lastused_tabpage = old_curtab;
 
-    entering_window(curwin);
+      entering_window(curwin);
+    }
 
     apply_autocmds(EVENT_WINNEW, NULL, NULL, false, curbuf);
-    apply_autocmds(EVENT_WINENTER, NULL, NULL, false, curbuf);
+    if (enter) {
+      apply_autocmds(EVENT_WINENTER, NULL, NULL, false, curbuf);
+    }
     apply_autocmds(EVENT_TABNEW, filename, filename, false, curbuf);
-    apply_autocmds(EVENT_TABENTER, NULL, NULL, false, curbuf);
+    if (enter) {
+      apply_autocmds(EVENT_TABENTER, NULL, NULL, false, curbuf);
+    }
```

Comment:
With `!enter`, these autocommands won't fire within the context of the new tabpage, which maybe they should. Also, they could free `old_curwin`.